File: src/sugtao4423/lod/IntentActivity.java
package sugtao4423.lod;

import java.util.HashMap;
import java.util.regex.Matcher;

import twitter4j.Status;
import twitter4j.TwitterException;
import android.annotation.TargetApi;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.DialogInterface.OnDismissListener;
import android.content.Intent;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import sugtao4423.lod.tweetlistview.TweetListAdapter;
import sugtao4423.lod.tweetlistview.TweetListView;
import sugtao4423.lod.userpage_fragment.UserPage;
import sugtao4423.lod.utils.Regex;

public class IntentActivity extends Activity{

	public static final String TWEET_ID = "tweetId";

	private App app;

	@Override
	protected void onCreate(Bundle savedInstanceState){
		super.onCreate(savedInstanceState);
		app = (App)getApplicationContext();

		if(!app.haveAccount()){
			startActivity(new Intent(this, StartOAuth.class));
			finish();
			return;
		}

		long tweetId = getIntent().getLongExtra(TWEET_ID, -1);
		if(tweetId == -1){
			jump();
		}else{
			showStatus(tweetId);
		}
	}

	public void jump(){
		if(Intent.ACTION_VIEW.equals(getIntent().getAction())){
			String uri = getIntent().getData().toString();
			Matcher status = Regex.statusUrl.matcher(uri);
			Matcher share = Regex.shareUrl.matcher(uri);
			Matcher user = Regex.userUrl.matcher(uri);
			if(status.find()){
				showStatus(Long.parseLong(status.group(2)));
			}else if(share.find()){
				Uri shareUri = Uri.parse(uri);
				HashMap<String, String> map = new HashMap<String, String>();
				for(String name : shareUri.getQueryParameterNames()){
					map.put(name, shareUri.getQueryParameter(name));
				}
				String text = (map.get("text") == null ? "" : (map.get("text") + " ")) +
						(map.get("url") == null ? "" : (map.get("url") + " ")) +
						(map.get("hashtags") == null ? "" : ("#" + map.get("hashtags").replace(",", " #") + " ")) +
						(map.get("via") == null ? "" : ("@" + map.get("via") + "さんから "));
				text = text.substring(0, text.length() - 1);
				Intent i = new Intent(IntentActivity.this, TweetActivity.class);
				i.putExtra(TweetActivity.INTENT_EXTRA_KEY_TYPE, TweetActivity.TYPE_EXTERNALTEXT);
				i.putExtra(TweetActivity.INTENT_EXTRA_KEY_TEXT, text);
				startActivity(i);
				finish();
			}else if(user.find()){
				Intent i = new Intent(IntentActivity.this, UserPage.class);
				i.putExtra(UserPage.INTENT_EXTRA_KEY_USER_SCREEN_NAME, user.group(2));
				startActivity(i);
				finish();
			}
		}else if(Intent.ACTION_SEND.equals(getIntent().getAction())){
			String subject = getIntent().getExtras().getString(Intent.EXTRA_SUBJECT);
			String text = getIntent().getExtras().getString(Intent.EXTRA_TEXT);
			text = (subject == null) ? text : (subject.isEmpty() ? text : (subject + " " + text));
			Intent i = new Intent(IntentActivity.this, TweetActivity.class);
			i.putExtra(TweetActivity.INTENT_EXTRA_KEY_TYPE, TweetActivity.TYPE_EXTERNALTEXT);
			i.putExtra(TweetActivity.INTENT_EXTRA_KEY_TEXT, text);
			startActivity(i);
			finish();
		}
	}

	public void showStatus(long tweetId){
		new AsyncTask<Long, Void, Status>(){
			@Override
			protected twitter4j.Status doInBackground(Long... params){
				try{
					return app.getTwitter().showStatus(params[0]);
				}catch(TwitterException e){
					return null;
				}
			}

			@TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR1)
			@Override
			protected void onPostExecute(twitter4j.Status status){
				if(status != null){
					TweetListView l = new TweetListView(IntentActivity.this);
					TweetListAdapter adapter = new TweetListAdapter(IntentActivity.this);
					adapter.add(status);
					l.setAdapter(adapter);
					adapter.setOnItemClickListener(new ListViewListener());
					adapter.setOnItemLongClickListener(new ListViewListener());
					AlertDialog.Builder builder = new AlertDialog.Builder(IntentActivity.this);
					builder.setView(l);
					builder.setOnDismissListener(new OnDismissListener(){
						@Override
						public void onDismiss(DialogInterface dialog){
							finish();
						}
					});
					AlertDialog dialog = builder.create();
					dialog.getWindow().setDimAmount(0f);
					dialog.show();
				}else{
					new ShowToast(getApplicationContext(), R.string.error_getStatus);
				}
			}
		}.execute(tweetId);
	}
}

File: app/src/main/java/sugtao4423/lod/IntentActivity.kt
package sugtao4423.lod

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import sugtao4423.lod.tweetlistview.TweetListAdapter
import sugtao4423.lod.tweetlistview.TweetListView
import sugtao4423.lod.ui.addaccount.AddAccountActivity
import sugtao4423.lod.ui.userpage.UserPageActivity
import sugtao4423.lod.utils.Regex
import twitter4j.TwitterException

class IntentActivity : AppCompatActivity() {

    companion object {
        const val TWEET_ID = "tweetId"
    }

    private lateinit var app: App

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        app = applicationContext as App

        if (!app.hasAccount) {
            startActivity(Intent(this, AddAccountActivity::class.java))
            finish()
            return
        }

        val tweetId = intent.getLongExtra(TWEET_ID, -1)
        if (tweetId == -1L) {
            jump()
        } else {
            showStatus(tweetId)
        }
    }

    private fun jump() {
        if (intent.action == Intent.ACTION_VIEW) {
            if (intent.data == null) {
                return
            }

            val uri = intent.data!!.toString()
            val status = Regex.statusUrl.matcher(uri)
            val share = Regex.shareUrl.matcher(uri)
            val user = Regex.userUrl.matcher(uri)
            when {
                status.find() -> {
                    showStatus(status.group(Regex.statusUrlStatusIdGroup)!!.toLong())
                }
                share.find() -> {
                    val shareUri = Uri.parse(uri)
                    val map = HashMap<String, String>()
                    shareUri.queryParameterNames.map {
                        map.put(it, shareUri.getQueryParameter(it) ?: "")
                    }
                    val text = arrayListOf<String>().apply {
                        map["text"]?.let {
                            add(it)
                        }
                        map["url"]?.let {
                            add(it)
                        }
                        map["hashtags"]?.let {
                            val str = "#" + it.replace(",", " #")
                            add(str)
                        }
                        map["via"]?.let {
                            val str = "@${it}さんから"
                            add(str)
                        }
                    }.joinToString(" ")
                    val i = Intent(this, TweetActivity::class.java).apply {
                        putExtra(TweetActivity.INTENT_EXTRA_KEY_TYPE, TweetActivity.TYPE_EXTERNALTEXT)
                        putExtra(TweetActivity.INTENT_EXTRA_KEY_TEXT, text)
                    }
                    startActivity(i)
                    finish()
                }
                user.find() -> {
                    val i = Intent(this, UserPageActivity::class.java)
                    i.putExtra(UserPageActivity.INTENT_EXTRA_KEY_USER_SCREEN_NAME, user.group(Regex.userUrlScreenNameGroup))
                    startActivity(i)
                    finish()
                }
            }
        } else if (intent.action == Intent.ACTION_SEND) {
            if (intent.extras == null) {
                return
            }

            val subject = intent.extras!!.getString(Intent.EXTRA_SUBJECT)
            val text = intent.extras!!.getString(Intent.EXTRA_TEXT)
            val tweetText = if (subject.isNullOrBlank()) {
                text
            } else {
                "$subject $text"
            }
            val i = Intent(this, TweetActivity::class.java).apply {
                putExtra(TweetActivity.INTENT_EXTRA_KEY_TYPE, TweetActivity.TYPE_EXTERNALTEXT)
                putExtra(TweetActivity.INTENT_EXTRA_KEY_TEXT, tweetText)
            }
            startActivity(i)
            finish()
        }
    }

    private fun showStatus(tweetId: Long) {
        CoroutineScope(Dispatchers.Main).launch {
            val result = withContext(Dispatchers.IO) {
                try {
                    app.twitter.showStatus(tweetId)
                } catch (e: TwitterException) {
                    null
                }
            }
            if (result != null) {
                val l = TweetListView(this@IntentActivity)
                val adapter = TweetListAdapter(this@IntentActivity)
                adapter.add(result)
                l.adapter = adapter
                AlertDialog.Builder(this@IntentActivity).apply {
                    setView(l)
                    setOnDismissListener {
                        finish()
                    }
                    window.setDimAmount(0f)
                    show()
                }
            } else {
                ShowToast(applicationContext, R.string.error_get_status)
            }
        }
    }

}
